File: internal/sheet/edit_test.go
package sheet

import (
	"testing"

	"github.com/flacyak/uno/internal/program"
)

// rows returns a fresh copy each call, since Set mutates in place.
func fixture() *Sheet {
	return New("sales.csv",
		[]string{"date", "region", "units"},
		[][]string{
			{"2026-07-01", "West", "1,204"},
			{"2026-07-01", "East", "987"},
			{"2026-07-02", "North", "1,455"},
		})
}

// An edit records what it replaced, which is what makes it readable on its own
// and reversible without re-reading the source.
func TestSetRecordsTheValueItReplaced(t *testing.T) {
	s := fixture()

	if err := s.Set(0, 2, "1204"); err != nil {
		t.Fatalf("Set: %v", err)
	}
	if got := s.Raw(0, 2); got != "1204" {
		t.Errorf("cell = %q, want %q", got, "1204")
	}

	log := s.Edits()
	if len(log) != 1 {
		t.Fatalf("log = %v, want one entry", log)
	}
	want := Edit{Seq: 1, Op: OpSet, Row: 0, Col: 2, Was: "1,204", Now: "1204"}
	if log[0] != want {
		t.Errorf("entry = %+v, want %+v", log[0], want)
	}
}

// Edits is handed to a worker goroutine that writes it while the person keeps
// typing, so it must not alias the log the sheet goes on appending to.
func TestEditsDoesNotAliasTheLog(t *testing.T) {
	s := fixture()
	if err := s.Set(0, 2, "1204"); err != nil {
		t.Fatalf("Set: %v", err)
	}

	snapshot := s.Edits()
	if err := s.Set(1, 2, "987!"); err != nil {
		t.Fatalf("Set: %v", err)
	}

	if len(snapshot) != 1 {
		t.Errorf("snapshot grew to %d entries", len(snapshot))
	}
	if s.EditCount() != 2 {
		t.Errorf("EditCount = %d, want 2", s.EditCount())
	}
}

// Fixing the last unparseable value in a column is what clears its warning
// badge, so the kind is re-inferred as the edit lands rather than at the next open.
func TestFixingTheLastBadValueClearsTheFlag(t *testing.T) {
	s := fixture()
	if c := s.Columns[2]; c.Kind != KindText || !c.Flagged {
		t.Fatalf("units started as %v flagged=%v, want a flagged text column",
			c.Kind, c.Flagged)
	}

	for _, e := range []struct {
		row int
		now string
	}{{0, "1204"}, {2, "1455"}} {
		if err := s.Set(e.row, 2, e.now); err != nil {
			t.Fatalf("Set: %v", err)
		}
	}

	if c := s.Columns[2]; c.Kind != KindNum || c.Flagged {
		t.Errorf("units is now %v flagged=%v, want an unflagged num column",
			c.Kind, c.Flagged)
	}
	if s.Columns[2].Header != "units" {
		t.Error("re-inferring the kind lost the column's header")
	}
}

// Ragged rows are normal in real exports. Editing a cell past the end of its
// row has to create it rather than be swallowed by At's tolerance of short rows.
func TestSetGrowsAShortRow(t *testing.T) {
	s := New("ragged.csv",
		[]string{"a", "b", "c"},
		[][]string{{"1"}, {"2", "3", "4"}})

	if err := s.Set(0, 2, "filled"); err != nil {
		t.Fatalf("Set: %v", err)
	}
	if got := s.Raw(0, 2); got != "filled" {
		t.Errorf("cell = %q, want %q", got, "filled")
	}
	if got := s.Raw(0, 1); got != "" {
		t.Errorf("the padded cell = %q, want it empty", got)
	}
}

// Replay is how a .uno rebuilds itself, and it must keep the log it replayed so
// that saving a reopened file preserves the history rather than starting a new one.
func TestReplayRebuildsAndKeepsTheLog(t *testing.T) {
	edits := []Edit{
		{Seq: 1, Op: OpSet, Row: 0, Col: 2, Was: "1,204", Now: "1204"},
		{Seq: 2, Op: OpSet, Row: 2, Col: 2, Was: "1,455", Now: "1455"},
	}

	s := fixture()
	if err := s.Replay(edits); err != nil {
		t.Fatalf("Replay: %v", err)
	}

	if got := s.Raw(0, 2); got != "1204" {
		t.Errorf("cell = %q, want the replayed value", got)
	}
	if s.EditCount() != 2 {
		t.Errorf("EditCount = %d, want the replayed log kept", s.EditCount())
	}
	if err := s.Set(1, 2, "986"); err != nil {
		t.Fatalf("Set: %v", err)
	}
	if got := s.Edits()[2].Seq; got != 3 {
		t.Errorf("seq after replay = %d, want it to carry on at 3", got)
	}
}

// A log that names a cell this sheet does not have is a log that does not belong
// to these bytes, and opening must say so rather than build a wrong grid.
func TestReplayRefusesALogThatDoesNotFit(t *testing.T) {
	for name, e := range map[string]Edit{
		"row past the end":    {Seq: 1, Op: OpSet, Row: 9, Col: 0, Now: "x"},
		"column past the end": {Seq: 1, Op: OpSet, Row: 0, Col: 9, Now: "x"},
		"unknown operation":   {Seq: 1, Op: "rule", Row: 0, Col: 0, Now: "x"},
	} {
		t.Run(name, func(t *testing.T) {
			if err := fixture().Replay([]Edit{e}); err == nil {
				t.Error("want an error, got nil")
			}
		})
	}
}

// prog parses a program the way a log line would, so a test that writes an
// unparseable one fails here rather than somewhere downstream.
func prog(t *testing.T, src string) program.Program {
	t.Helper()
	p, err := program.Parse(src)
	if err != nil {
		t.Fatalf("Parse(%q): %v", src, err)
	}
	return p
}

// The point of a column op is that the log grows with what a person did and not
// with how much data they did it to. Three cells change and one line is written.
func TestApplyWritesOneOperationForAWholeColumn(t *testing.T) {
	s := fixture()

	if err := s.Apply(2, prog(t, `replace(/,/, "")`)); err != nil {
		t.Fatalf("Apply: %v", err)
	}

	for row, want := range map[int]string{0: "1204", 1: "987", 2: "1455"} {
		if got := s.Raw(row, 2); got != want {
			t.Errorf("cell (%d,2) = %q, want %q", row, got, want)
		}
	}

	log := s.Edits()
	if len(log) != 1 {
		t.Fatalf("log = %v, want one entry", log)
	}
	want := Edit{Seq: 1, Op: OpApply, Row: NoRow, Col: 2, Now: `replace(/,/, "")`}
	if log[0] != want {
		t.Errorf("entry = %+v, want %+v", log[0], want)
	}
}

// Fixing a whole column is what clears its warning badge, and the kind has to be
// re-read as the operation lands rather than at the next open.
func TestApplyingAProgramRenamesTheColumn(t *testing.T) {
	s := fixture()
	if c := s.Columns[2]; c.Kind != KindText || !c.Flagged {
		t.Fatalf("units = %v flagged=%v, want the warning badge to start on", c.Kind, c.Flagged)
	}

	if err := s.Apply(2, prog(t, `replace(/,/, "")`)); err != nil {
		t.Fatalf("Apply: %v", err)
	}
	if c := s.Columns[2]; c.Kind != KindNum || c.Flagged {
		t.Errorf("units = %v flagged=%v, want num and unflagged", c.Kind, c.Flagged)
	}
}

// A transform rewrites values that are there. A row that never had this column
// has no value to be wrong about, and inventing an empty cell would change the
// shape of the data on the strength of an inference.
func TestApplySkipsRowsWithoutTheColumn(t *testing.T) {
	s := New("ragged.csv",
		[]string{"date", "region", "units"},
		[][]string{
			{"2026-07-01", "West", "1,204"},
			{"2026-07-01"},
		})

	if err := s.Apply(2, prog(t, `replace(/,/, "")`)); err != nil {
		t.Fatalf("Apply: %v", err)
	}
	if got := s.Raw(0, 2); got != "1204" {
		t.Errorf("cell (0,2) = %q, want %q", got, "1204")
	}
	if got := s.Rows(); got != 2 {
		t.Errorf("Rows = %d, want 2", got)
	}
	if got := s.Raw(1, 2); got != "" {
		t.Errorf("short row grew a cell holding %q", got)
	}
}

// Undo replays, and a column op is the case that mechanism exists for: there is
// no old value to put back, so the rebuild has to produce the same column the
// operation did the first time.
func TestReplayRebuildsAnAppliedColumn(t *testing.T) {
	s := fixture()
	if err := s.Set(1, 2, "9,870"); err != nil {
		t.Fatalf("Set: %v", err)
	}
	if err := s.Apply(2, prog(t, `replace(/,/, "")`)); err != nil {
		t.Fatalf("Apply: %v", err)
	}
	log := s.Edits()

	rebuilt := fixture()
	if err := rebuilt.Replay(log); err != nil {
		t.Fatalf("Replay: %v", err)
	}

	for row := 0; row < 3; row++ {
		if got, want := rebuilt.Raw(row, 2), s.Raw(row, 2); got != want {
			t.Errorf("cell (%d,2) = %q, want %q", row, got, want)
		}
	}
	if c := rebuilt.Columns[2]; c.Kind != KindNum || c.Flagged {
		t.Errorf("units = %v flagged=%v, want num and unflagged", c.Kind, c.Flagged)
	}
}

// A log naming a program this build cannot read has to fail before a single cell
// moves. Refusing to open a workspace is recoverable; half-transforming one is
// not.
func TestReplayRefusesAProgramItCannotRead(t *testing.T) {
	s := fixture()
	err := s.Replay([]Edit{{Seq: 1, Op: OpApply, Row: NoRow, Col: 2, Now: "explode()"}})
	if err == nil {
		t.Fatal("Replay accepted a program it cannot run")
	}
	if got := s.Raw(0, 2); got != "1,204" {
		t.Errorf("cell (0,2) = %q, want it untouched at %q", got, "1,204")
	}
}

File: internal/grid/tests/sheet/edit.test.ts
import { describe, expect, test } from "vite-plus/test";

import { NO_ROW, Op, Sheet } from "../../src/sheet/index.ts";
import type { Edit } from "../../src/sheet/index.ts";
import { parse as parseProgram } from "../../src/program/index.ts";

const REGION = 1;
const UNITS = 2;

// A fresh sheet each call, since the operations mutate in place.
function fixture(): Sheet {
  return new Sheet(
    "sales.csv",
    ["date", "region", "units"],
    [
      ["2026-07-01", "West", "1,204"],
      ["2026-07-01", "East", "987"],
      ["2026-07-02", "North", "1,455"],
    ],
  );
}

// An edit records what it replaced, which is what makes it readable on its own
// and reversible without re-reading the source.
test("set records the value it replaced", () => {
  const s = fixture();
  s.set(0, UNITS, "1204");

  expect(s.raw(0, UNITS)).toBe("1204");

  const log = s.edits();
  expect(log).toHaveLength(1);
  expect(log[0]).toEqual({ seq: 1, op: Op.Set, row: 0, col: UNITS, was: "1,204", now: "1204" });
});

// edits() is handed to whatever writes it while the person keeps typing, so it
// must not alias the log the sheet goes on appending to.
test("edits does not alias the log", () => {
  const s = fixture();
  s.set(0, UNITS, "1204");

  const snapshot = s.edits();
  s.set(1, UNITS, "987!");

  expect(snapshot).toHaveLength(1);
  expect(s.editCount()).toBe(2);
});

// Fixing the last unparseable value in a column is what clears its warning
// badge, so the kind is re-inferred as the edit lands rather than at the next
// open.
test("fixing the last bad value clears the flag", () => {
  const s = fixture();
  expect(s.columns[UNITS]!.kind).toBe("text");
  expect(s.columns[UNITS]!.flagged).toBe(true);

  s.set(0, UNITS, "1204");
  s.set(2, UNITS, "1455");

  expect(s.columns[UNITS]!.kind).toBe("num");
  expect(s.columns[UNITS]!.flagged).toBe(false);
  expect(s.columns[UNITS]!.header, "re-inferring lost the header").toBe("units");
});

// Ragged rows are normal in real exports. Editing a cell past the end of its
// row has to create it rather than be swallowed by raw's tolerance of short rows.
test("set grows a short row", () => {
  const s = new Sheet("ragged.csv", ["a", "b", "c"], [["1"], ["2", "3", "4"]]);

  s.set(0, 2, "filled");

  expect(s.raw(0, 2)).toBe("filled");
  expect(s.raw(0, 1), "the padded cell").toBe("");
});

// Replay is how a .uno rebuilds itself, and it must keep the log it replayed so
// that saving a reopened file preserves the history rather than starting a new
// one.
test("replay rebuilds and keeps the log", () => {
  const edits: Edit[] = [
    { seq: 1, op: Op.Set, row: 0, col: UNITS, was: "1,204", now: "1204" },
    { seq: 2, op: Op.Set, row: 2, col: UNITS, was: "1,455", now: "1455" },
  ];

  const s = fixture();
  s.replay(edits);

  expect(s.raw(0, UNITS)).toBe("1204");
  expect(s.editCount()).toBe(2);

  s.set(1, UNITS, "986");
  expect(s.edits()[2]!.seq, "seq carries on after a replay").toBe(3);
});

// A log that names a cell this sheet does not have is a log that does not
// belong to these bytes, and opening must say so rather than build a wrong grid.
describe("replay refuses a log that does not fit", () => {
  const cases: Array<[string, Edit]> = [
    ["row past the end", { seq: 1, op: Op.Set, row: 9, col: 0, now: "x" }],
    ["column past the end", { seq: 1, op: Op.Set, row: 0, col: 9, now: "x" }],
    ["unknown operation", { seq: 1, op: "rule" as Op, row: 0, col: 0, now: "x" }],
  ];

  for (const [name, e] of cases) {
    test(name, () => {
      expect(() => fixture().replay([e])).toThrow();
    });
  }
});

// The point of a column op is that the log grows with what a person did and not
// with how much data they did it to. Three cells change and one line is written.
test("apply writes one operation for a whole column", () => {
  const s = fixture();
  s.apply(UNITS, parseProgram('replace(/,/, "")'));

  expect(s.raw(0, UNITS)).toBe("1204");
  expect(s.raw(1, UNITS)).toBe("987");
  expect(s.raw(2, UNITS)).toBe("1455");

  const log = s.edits();
  expect(log).toHaveLength(1);
  expect(log[0]).toEqual({
    seq: 1,
    op: Op.Apply,
    row: NO_ROW,
    col: UNITS,
    now: 'replace(/,/, "")',
  });
});

// Fixing a whole column is what clears its warning badge, and the kind has to
// be re-read as the operation lands rather than at the next open.
test("applying a program renames the column", () => {
  const s = fixture();
  expect(s.columns[UNITS]!.flagged).toBe(true);

  s.apply(UNITS, parseProgram('replace(/,/, "")'));

  expect(s.columns[UNITS]!.kind).toBe("num");
  expect(s.columns[UNITS]!.flagged).toBe(false);
});

// A transform rewrites values that are there. A row that never had this column
// has no value to be wrong about, and inventing an empty cell would change the
// shape of the data on the strength of an inference.
test("apply skips rows without the column", () => {
  const s = new Sheet(
    "ragged.csv",
    ["date", "region", "units"],
    [["2026-07-01", "West", "1,204"], ["2026-07-01"]],
  );

  s.apply(UNITS, parseProgram('replace(/,/, "")'));

  expect(s.raw(0, UNITS)).toBe("1204");
  expect(s.rows()).toBe(2);
  expect(s.raw(1, UNITS), "the short row grew a cell").toBe("");
});

// Undo replays, and a column op is the case that mechanism exists for: there is
// no old value to put back, so the rebuild has to produce the same column the
// operation did the first time.
test("replay rebuilds an applied column", () => {
  const s = fixture();
  s.set(1, UNITS, "9,870");
  s.apply(UNITS, parseProgram('replace(/,/, "")'));

  const rebuilt = fixture();
  rebuilt.replay(s.edits());

  for (let row = 0; row < s.rows(); row++) {
    expect(rebuilt.raw(row, UNITS), `row ${row}`).toBe(s.raw(row, UNITS));
  }
  expect(rebuilt.columns[UNITS]!.kind).toBe("num");
  expect(rebuilt.columns[UNITS]!.flagged).toBe(false);
});

// The recogniser learns a program from cells a person fixed, and those were
// typed before the apply that follows. Running a program that is not idempotent
// over them again would fix them twice: West-q3 would read West-q3-q3.
test("apply leaves a cell it would have fixed the same way, and rewrites one it would not", () => {
  const s = fixture();
  s.set(0, REGION, "West-q3");
  s.set(1, REGION, "Eas");
  s.set(1, REGION, "East-q3"); // a slip on the way is still East to East-q3
  s.set(2, REGION, "N");
  s.apply(REGION, parseProgram('concat(slice(0, len), "-q3")'));

  expect(s.raw(0, REGION)).toBe("West-q3");
  expect(s.raw(1, REGION)).toBe("East-q3");
  expect(s.raw(2, REGION), "N is not what the program makes of North").toBe("N-q3");

  const rebuilt = fixture();
  rebuilt.replay(s.edits());
  for (let row = 0; row < s.rows(); row++) {
    expect(rebuilt.raw(row, REGION), `row ${row} after replay`).toBe(s.raw(row, REGION));
  }
});

// A log naming a program this build cannot read has to fail before a single
// cell moves. Refusing to open a workspace is recoverable; half-transforming
// one is not.
test("replay refuses a program it cannot read", () => {
  const s = fixture();
  expect(() =>
    s.replay([{ seq: 1, op: Op.Apply, row: NO_ROW, col: UNITS, now: "explode()" }]),
  ).toThrow();
  expect(s.raw(0, UNITS), "the cell should be untouched").toBe("1,204");
});
